x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -694,6 +715,31 @@ static PreAnalysisResult fieldNames(LogicalPlan parsed, Set<String> enrichPolicy
         }
     }
 
+    /**
+     * Could a plan "accidentally" override aliases?
+     * Examples are JOIN and ENRICH, that _could_ produce fields with the same
+     * name of an existing alias, based on their index mapping.
+     * Here we just have to consider commands where this information is not available before index resolution,
+     * eg. EVAL, GROK, DISSECT can override an alias, but we know it in advance, ie. we don't need to resolve indices to know.
+     */
+    private static boolean couldOverrideAliases(LogicalPlan p) {
+        return (p instanceof Aggregate
```

Comment:
Why are you listing all of these that **cannot** define an additional "hidden" Attribute instead of checking those that **can**?

Reply:
To avoid that it breaks when we add new commands that behave like JOIN.
As a follow-up, IMHO we should define an interface for commands that allow this optimization (similar to `SortAgnostic`) and replace this long list with a single `instanceof`

Reply:
If we add a command that doesn't behave like JOIN, then it could break the other way around. Imo, it makes much more sense to list those commands that are "special".

Reply:
> If we add a command that doesn't behave like JOIN, then it could break the other way around

This is interesting, I'm not sure I have completely clear the implications of this aspect. 

If I got it right, we use these field names to limit the scope of the field_caps; if we add a field that does not exist in any of the involved indices, will field_caps fail? 

My understanding is that it will work fine, just ignoring the additional fields.

If it wasn't the case, then we would be in a catch-22: we couldn't know which fields to send before validating the query, and we couldn't validate the query before sending the fields to field_caps.

The fact that everything works fine makes me think that it's safe, but maybe I'm missing something.

Reply:
How about `completion` command (I see it implements `GeneratingPlan`), is this a command that can add a "hidden" attribute that can be overiden?

Reply:
I think Completion is safe (the target attribute is defined at parsing time), but I have to double-check.